api/api_ethereum.go

```diff
@@ -460,8 +460,10 @@ func (api *EthereumAPI) GetHeaderByHash(ctx context.Context, hash common.Hash) m
 // * When fullTx is true all transactions in the block are returned, otherwise
 //   only the transaction hash is returned.
 func (api *EthereumAPI) GetBlockByNumber(ctx context.Context, number rpc.BlockNumber, fullTx bool) (map[string]interface{}, error) {
-	klaytnBlock, err := api.publicBlockChainAPI.b.BlockByNumber(ctx, number)
-	if klaytnBlock != nil && err == nil {
+	// Klaytn backend returns error when there is no matched block but
+	// Ethereum returns it as nil without error, so we should return is as nil when there is no matched block.
+	klaytnBlock, _ := api.publicBlockChainAPI.b.BlockByNumber(ctx, number)
+	if klaytnBlock != nil {
```

Comment:
What about this?

```go
klaytnBlock, err := api.publicBlockChainAPI.b.BlockByNumber(ctx, number)
if err != nil {
  if err == errBlockNotFound {
    return nil, nil
  }
  return nil, err
}
		response, err := api.rpcMarshalBlock(klaytnBlock, true, fullTx)
		if err == nil && number == rpc.PendingBlockNumber {
			// Pending blocks need to nil out a few fields
			for _, field := range []string{"hash", "nonce", "miner"} {
				response[field] = nil
			}
		}
		return response, err

```

Reply:
@kjhman21 
Looks great to me.
I'll update this PR.

Reply:
@kjhman21 

PTAL [this commit](https://github.com/klaytn/klaytn/pull/1261/commits/b04e32c4267ae4150306cdde91551f4bcbf3be76).

I considered defining not existing err as separate error variables in cn package and try to pack with that error like below, but there is a import cycle problem between cn package and api package.
```
// node/cn/api_backend.go
var ErrBlockNotFound = errors.New("the block does not exist")

func (b *CNAPIBackend) BlockByHash(ctx context.Context, hash common.Hash) (*types.Block, error) {
	block := b.cn.blockchain.GetBlockByHash(hash)
	if block == nil {
		return nil, fmt.Errorf("%w (block hash: %s)", ErrBlockNotFound, hash.String())
	}
	return block, nil
}

// api/api_ethereum.go
func (api *EthereumAPI) GetBlockByNumber(ctx context.Context, number rpc.BlockNumber) (map[string]interface{}, error) {
	klaytnHeader, err := api.publicBlockChainAPI.b.BlockByNumber(ctx, number)
	if err != nil {
		if errors.Is(err, cn.ErrBlockNotFound) {
			return nil, nil
		}
		return nil, err
	}
        // ...
}
```

So I checked the error whether it is containing "does not exist" text or not in its error message.